app/routes/thematic/geoencoder.py

```diff
@@ -0,0 +1,216 @@
+import re
+from typing import Any, Dict, List, Optional
+
+from fastapi import APIRouter, HTTPException, Query
+from unidecode import unidecode
+
+from app.crud.versions import get_version, get_version_names
+from app.errors import RecordNotFoundError
+from app.models.pydantic.responses import Response
+from app.routes import VERSION_REGEX
+from app.routes.datasets.queries import _query_dataset_json
+
+router = APIRouter()
+
+
+@router.get(
+    "/geoencode",
+    tags=["Geoencoder"],
+    status_code=200,
+)
+async def geoencode(
+    *,
+    admin_source: str = Query(
+        "GADM", description="The source of administrative boundaries to use."
+    ),
+    admin_version: str = Query(
+        ...,
+        description="Version of the administrative boundaries dataset to use.",
+    ),
+    country: str = Query(
+        description="Name of the country to match.",
+    ),
+    region: Optional[str] = Query(
+        None,
+        description="Name of the region to match.",
+    ),
+    subregion: Optional[str] = Query(
+        None,
+        description="Name of the subregion to match.",
+    ),
+    normalize_search: bool = Query(
+        True,
+        description="Whether or not to perform a case- and accent-insensitive search.",
+    ),
+):
+    """Look up administrative boundary IDs matching a specified country name
+    (and region name and subregion names, if specified).
+    """
+    admin_source_to_dataset: Dict[str, str] = {"GADM": "gadm_administrative_boundaries"}
+
+    try:
+        dataset: str = admin_source_to_dataset[admin_source.upper()]
+    except KeyError:
+        raise HTTPException(
+            status_code=400,
+            detail=(
+                "Invalid admin boundary source. Valid sources:"
+                f" {[source for source in admin_source_to_dataset.keys()]}"
+            ),
+        )
+
+    version_str: str = "v" + str(admin_version).lstrip("v")
+
+    await version_is_valid(dataset, version_str)
```

Comment:
Like mentioned about the documentation above, should it be well known in advance which versions we support for providers, rather than just trying and throwing back an error?

Reply:
I fundamentally agree that your suggestion is a good one, but it turns out to be difficult to do in practice. Especially considering the limited use of this endpoint and the fact that we will be omitting it from the docs.